From a46ed472afb71738f44bc93e2e7abdc9590f0daf Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Thu, 6 Jun 2024 15:32:39 +0200
Subject: [PATCH 622/954] feat(stm32mp1): update STPMIC1 driver to handle more
 properties

Update STPMIC1 driver to handle more properties.
The goal if to simplify plat ddr initialization.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 1bb3b4f1bff5263f5c085369bfa561fc32f787ce feat(stm32mp1): update STPMIC1 driver to handle more properties]
Change-Id: Ibc08d86e301fa8bdbf82876f7ccb6cee85f38a96
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/389335
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/pmic/stm32mp_pmic.c | 44 ++++++++++++++++++++++++++++------
 1 file changed, 37 insertions(+), 7 deletions(-)

diff --git a/drivers/st/pmic/stm32mp_pmic.c b/drivers/st/pmic/stm32mp_pmic.c
index 0e774b848c..9db192ebad 100644
--- a/drivers/st/pmic/stm32mp_pmic.c
+++ b/drivers/st/pmic/stm32mp_pmic.c
@@ -382,6 +382,24 @@ static const struct regul_description pmic_regs[NB_REG] = {
 	[STPMIC1_SW_OUT] = DEFINE_REGU("pwr_sw2"),
 };
 
+static int handle_pmic_property(void *fdt, int subnode,
+				const struct regul_description *desc,
+				const char *property, uint16_t flag)
+{
+	if (fdt_getprop(fdt, subnode, property, NULL)  != NULL) {
+		int ret;
+
+		VERBOSE("%s: %s\n", desc->node_name, property);
+		ret = pmic_set_flag(desc, flag);
+		if (ret != 0) {
+			ERROR("set %s failed\n", property);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
 static int register_pmic(void)
 {
 	void *fdt;
@@ -424,13 +442,25 @@ static int register_pmic(void)
 			return ret;
 		}
 
-		if (fdt_getprop(fdt, subnode, "st,mask-reset", NULL)  != NULL) {
-			VERBOSE("%s: set mask-reset\n", desc->node_name);
-			ret = pmic_set_flag(desc, REGUL_MASK_RESET);
-			if (ret != 0) {
-				ERROR("set mask-reset failed\n");
-				return ret;
-			}
+		ret = handle_pmic_property(fdt, subnode, desc,
+					   "st,mask-reset",
+					   REGUL_MASK_RESET);
+		if (ret != 0) {
+			return ret;
+		}
+
+		ret = handle_pmic_property(fdt, subnode, desc,
+					   "st,regulator-sink-source",
+					   REGUL_SINK_SOURCE);
+		if (ret != 0) {
+			return ret;
+		}
+
+		ret = handle_pmic_property(fdt, subnode, desc,
+					   "st,regulator-bypass",
+					   REGUL_ENABLE_BYPASS);
+		if (ret != 0) {
+			return ret;
 		}
 	}
 
-- 
2.53.0

